scoutcloud/scoutcloud/src/logic/users/handlers/crud.rs

```diff
@@ -1,11 +1,56 @@
 use crate::{
-    logic::{AuthError, UserToken},
+    logic::{users::register_promo::try_find_promo, AuthError, UserToken},
     server::proto,
 };
-use sea_orm::{prelude::*, ConnectionTrait, DatabaseConnection, DbErr, QueryOrder, QuerySelect};
-
+use sea_orm::{
+    prelude::*, ActiveValue::Set, ConnectionTrait, DatabaseConnection, DbErr, NotSet, QueryOrder,
+    QuerySelect, TransactionTrait,
+};
 const RECENT_ACTIONS_LIMIT: u64 = 50;
 
+pub async fn register_profile(
+    db: &DatabaseConnection,
+    profile: &proto::RegisterProfileRequestInternal,
+) -> Result<proto::RegisterProfileResponseInternal, AuthError> {
+    let maybe_promo = if let Some(promo) = &profile.promo {
+        Some(try_find_promo(db, promo).await?)
+    } else {
+        None
+    };
+    let tx = db.begin().await?;
+    let user = scoutcloud_entity::users::ActiveModel {
+        project_title: Set(Some(profile.project_title.to_string())),
+        email: Set(profile.email.to_string()),
+        max_instances: maybe_promo
+            .as_ref()
+            .map(|p| Set(p.user_max_instances))
+            .unwrap_or(NotSet),
+        ..Default::default()
+    }
+    .insert(&tx)
+    .await?;
+    if let Some(promo) = maybe_promo {
+        scoutcloud_entity::balance_changes::ActiveModel {
+            user_id: Set(user.id),
+            amount: Set(promo.user_initial_balance),
+            note: Set(Some(format!("From promo code: {}", promo.code))),
+            register_promo_id: Set(Some(promo.id)),
+            ..Default::default()
+        }
+        .insert(&tx)
+        .await?;
+    }
+    let token = issue_auth_token(&tx, user.id, "initial token").await?;
+    tx.commit().await?;
+    let user_token = UserToken::try_from_token_value(db, &token.token_value).await?;
```

Comment:
As I understand it, that should always succeed. But in case if it fails, now we return `Authenticated` to the users, while actually that is internal error